# Chapter 2: Overview of Data Mining Process


> (c) 2019-2020 Galit Shmueli, Peter C. Bruce, Peter Gedeck 
>
> _Data Mining for Business Analytics: Concepts, Techniques, and Applications in Python_ (First Edition) 
> Galit Shmueli, Peter C. Bruce, Peter Gedeck, and Nitin R. Patel. 2019.
>
> Date: 2020-03-08
>
> Python Version: 3.8.2
> Jupyter Notebook Version: 5.6.1
>
> Packages:
>   - pandas: 1.0.1
>   - scikit-learn: 0.22.2
>
> The assistance from Mr. Kuber Deokar and Ms. Anuja Kulkarni in preparing these solutions is gratefully acknowledged.


In [1]:
# import required functionality for this chapter
from pathlib import Path

import pandas as pd
from sklearn.preprocessing import StandardScaler

In [2]:
# Working directory:
#
# We assume that data are kept in the same directory as the notebook. If you keep your 
# data in a different folder, replace the argument of the `Path`
DATA = Path('.').resolve().parent / 'data'
FIGURES = Path('.').resolve().parent / 'figures' / 'chapter_02'
FIGURES.mkdir(exist_ok=True, parents=True)
# and then load data using 
#
# pd.read_csv(DATA / ‘filename.csv’)

# Problem 2.1 Supervised or Unsupervised Task

Assuming that data mining techniques are to be used in the following cases, identify whether the task required is supervised or unsupervised learning.

__2.1.a.__ Deciding whether to issue a loan to an applicant based on demographic and financial data (with reference to a database of similar data on prior customers).

__Answer:__ This is supervised learning, because the database includes information on whether the loan was approved or not.

__2.1.b.__ In an online bookstore, making recommendations to customers concerning additional items to buy based on the buying patterns in prior transactions.

__Answer:__ This is unsupervised learning, because there is no apparent outcome (e.g., whether the recommendation was adopted or not).

__2.1.c.__ Identifying a network data packet as dangerous (virus, hacker attack) based on comparison to other packets whose threat status is known.

__Answer:__ This is supervised learning, because for the other packets the status is known.

__2.1.d.__ Identifying segments of similar customers.

__Answer:__ This is unsupervised learning because there is no known outcome (though once you use unsupervised learning to identify segments, you could use supervised learning to classify new customers into those segments).

__2.1.e.__ Predicting whether a company will go bankrupt based on comparing its financial data to those of similar bankrupt and nonbankrupt firms.

__Answer:__ This is supervised learning, because the status of the similar firms is known.

__2.1.f.__ Estimating the repair time required for an aircraft based on a trouble ticket.

__Answer:__ This is supervised learning, because there is likely to be knowledge of actual (historic) repair times of similar repairs.

__2.1.g.__ Automated sorting of mail by zip code scanning.

__Answer:__ This is supervised learning, as there is likely to be knowledge about whether the sorting was correct in previous mail sorting.

__2.1.h.__ Printing of custom discount coupons at the conclusion of a grocery store checkout based on what you just bought and what others have bought previously.

__Answer:__ This is unsupervised learning, if we assume that we do not know what will be purchased in the future.


# Problem 2.2 

Describe the difference in roles assumed by the validation partition and the test partition.

__Answer:__ 
The validation partition is used to assess the performance of each supervised learning model so that we can compare models and pick the best one. In some algorithms (e.g., classification and regression trees, k-nearest neighbors) the validation partition may be used in automated fashion to tune and improve the model. This means that the validation data are actually used to help build the model. 

The test data partition is used for assessing the performance of the final chosen model on new data. The test data are not used to compare models, or to further tweak the model or improve its fit. (If the test data were used for these purposes, they would play a role in building or selecting the best model, and would no longer provide an unbiased assessment of the chosen model's performance with completely new data.)

# Problem 2.3 
Consider the sample from a database of credit applicants in Table 2.16. Comment on the likelihood that it was sampled randomly, and whether it is likely to be a useful sample.

![Table2.16.PNG](img/Table2.16.PNG)

__Answer:__ This sample is not selected randomly as we can see from “observation #”, that there is pattern in the observations chosen for the sample. In particular, every 8th observation from the database was selected for the sample. When we select data with such a pre-decided methodology it might introduce a bias in the selected data set. This is true when the order of the observations in the dataset has some meaning (e.g., chronological order).

# Problem 2.4 
Consider the sample from a bank database shown in Table 2.17; it was selected randomly from a larger database to be the training set. _Personal Loan_ indicates whether a solicitation for a personal loan was accepted and is the response variable. A campaign is planned for a similar solicitation in the future and the bank is looking for a model that will identify likely responders. Examine the data carefully and indicate what your next step would be. 


![Table2.17.PNG](img/Table2.17.PNG)


__Answer:__
Since there are only 18 records and 9 predictor variables in the sample, the next step before building a model is to take a larger sample.  18 records is too few to support a model that considers 9 predictors.  How big a sample? The availability of data, the cost and effort involved in data handling, and software capabilities are the main constraining factors.  Also, it is useful to check the number of available responses, and the response ratio, in the larger database.  If the response ratio is very low, it would be worthwhile to oversample the cases where response is positive (or, in other words, undersample the non-response cases).  Two other issues:

- Zip code should probably be aggregated at a higher level than 5 digits, which would likely produce an unmanageable number of predictor variables.  For example, it could be aggregated at the level of the first 3 digits.
- It also seems that the key information in “Mortgage” is whether the person has a mortgage, and not so much the level of the mortgage, so some consideration could be given to converting this to a binary variable.

See also the 2016 blog post by  Tom Fawcett: [Learning from Imbalanced Classes](http://www.kdnuggets.com/2016/08/learning-from-imbalanced-classes.html/) 

# Problem 2.5
Using the concept of overfitting, explain why when a model is fit to training data, zero error with those data is not necessarily good.

__Answer:__
Overfitting occurs when the model captures not only the generalizable pattern in the data, but also the error. When we split the data into training and validation sets, we assume that the same pattern (if there is a pattern) exists in both, and that they differ only in the error that they contain. An absurd and false model may fit perfectly (on training data set) if the model has enough complexity. Therefore we may get zero error for such a model using the training dataset. Such a model, however, is not likely to give useful results on the validation data set.

# Problem 2.6 
In fitting a model to classify prospects as purchasers or nonpurchasers, a certain company drew the training data from internal data that include demographic and purchase information. Future data to be classified will be lists purchased from other sources,
with demographic (but not purchase) data included. It was found that “refund issued” was a useful predictor in the training data. Why is this not an appropriate variable to include in the model?

__Answer:__
The variable “refund issued” is unknown prior to the actual purchase, and therefore is not useful in a predictive model of future purchase behavior. In fact, “refund issued” can only be present for actual purchases but never for non-purchases. This explains why it was found to be closely related to purchase/non-purchase.

# Problem 2.7
A dataset has 1000 records and 50 variables with 5% of the values missing, spread randomly throughout the records and variables. An analyst decides to remove records with missing values. About how many records would you expect to be removed? 

__Answer:__ 
For a record to have all values present, it must avoid having a missing value (P = 0.95) for each of 50 records. The chance that a given record will escape having a missing value for two variables is 0.95 * 0.95 = 0.903. The chance that a given record would escape having a missing value for all 50 records is (0.95)^50 = 0.076945. This implies that 1-0.076944 = 0.9231 (92.31%) of all records will have missing values and would be deleted.

# Problem 2.8
Normalize the data in Table 2.18, showing calculations.

![Table2.18.PNG](img/Table2.18.PNG)

__Answer:__
Normalization of a measurement is obtained by subtracting the (column) average from each measurement and dividing the difference by the (column) standard deviation.

For variable Age (years):
Mean =44.66667 and Standard deviation (std) = 14.97554

For variable Income ($):
Mean=98.66667 and Standard deviation (std) = 62.86706

For normalizing age for observation # 1 (Here age = 25):
After subtracting the average and dividing by standard deviation, the normalized age = -1.438596.

For normalizing income for observation # 1 (Here Income = 49000):
After subtracting the average income and dividing by standard deviation, the normalized income = -0.865431.

Let's normalize the data using sklearn's preprocessing 


In [3]:
# import the required functionality for this problem
import numpy as np
import pandas as pd
from sklearn import preprocessing

# create a data frame
df = pd.DataFrame({
    'Age': [25, 56, 65, 32, 41, 49],
    'Income': [49000, 156000, 99000, 192000, 39000, 57000]
})
print(df)

   Age  Income
0   25   49000
1   56  156000
2   65   99000
3   32  192000
4   41   39000
5   49   57000


In [4]:
# normalize the data
scaler = StandardScaler()
df_norm = pd.DataFrame(scaler.fit_transform(df), index=df.index, columns=df.columns)
print(df_norm)

        Age    Income
0 -1.438597 -0.865431
1  0.829022  0.999021
2  1.487363  0.005808
3 -0.926554  1.626314
4 -0.268213 -1.039679
5  0.316979 -0.726033


# Problem 2.9
Statistical distance between records can be measured in several ways. Consider Euclidean distance, measured as the square root of the sum of the squared differences. For the first two records in Table 2.17, it is

\begin{equation*}
\sqrt{(25-56)^2 + (49000-156000)^2}
\end{equation*}

Can normalizing the data change which two records are farthest from each other in terms of Euclidean distance?

__Answer:__
Yes, it can. By normalizing we equate the scales of the different variables, and therefore the Euclidean distance can dramatically change. In the example age is in single years while income is in thousands of dollars. The Euclidean distance on the raw data is therefore almost completely determined by income and unaffected by age. In contrast, after normalizing age and income will be on the same scale. Age will then have a much larger impact on the Euclidean distance. To see this in practice, examine the table below that compare the Euclidean distance before and after normalizing the data.

We now see that records 4 and 5 (distance = 153000) are farthest from each other before normalizing, whereas records 1 and 3 are farthest from each other after normalizing.

In [5]:
# euclidean distances for original data
from sklearn.metrics import pairwise
d = pairwise.pairwise_distances(df, metric='euclidean')
pd.DataFrame(d, columns=df.index, index=df.index)

0              1             2              3              4  \
0       0.000000  107000.004491  50000.016000  143000.000171   10000.012800   
1  107000.004491       0.000000  57000.000711   36000.008000  117000.000962   
2   50000.016000   57000.000711      0.000000   93000.005855   60000.004800   
3  143000.000171   36000.008000  93000.005855       0.000000  153000.000265   
4   10000.012800  117000.000962  60000.004800  153000.000265       0.000000   
5    8000.036000   99000.000247  42000.003048  135000.001070   18000.001778   

               5  
0    8000.036000  
1   99000.000247  
2   42000.003048  
3  135000.001070  
4   18000.001778  
5       0.000000

In [6]:
# euclidean distances for normalized data
d_norm = pairwise.pairwise_distances(df_norm, metric='euclidean')
pd.DataFrame(d_norm, columns=df_norm.index, index=df_norm.index)

0         1         2         3             4         5
0  0.000000  2.935690  3.052916  2.543812  1.183284e+00  1.761101
1  2.935690  0.000000  1.191589  1.864280  2.315215e+00  1.799444
2  3.052916  1.191589  0.000000  2.907409  2.043303e+00  1.380358
3  2.543812  1.864280  2.907409  0.000000  2.746075e+00  2.660809
4  1.183284  2.315215  2.043303  2.746075  2.107342e-08  0.663945
5  1.761101  1.799444  1.380358  2.660809  6.639453e-01  0.000000

# Problem 2.10
Two models are applied to a dataset that has been partitioned. Model A is considerably more accurate than model B on the training data, but slightly less accurate than model B on the validation data. Which model are you more likely to consider for final deployment?

__Answer:__
We prefer the model with the lowest error on the validation data. Model A might be overfitting the training data. We would therefore select model B for deployment on new data.

# Problem 2.11
The dataset *ToyotaCorolla.csv* contains data on used cars on sale during the late summer of 2004 in the Netherlands. It has 1436 records containing details on 38 attributes, including *Price, Age, Kilometers, HP,* and other specifications. 

We plan to analyze the data using various data mining techniques described in future chapters. Prepare the data for use as follows:

__2.11.a.__ The dataset has two categorical attributes, Fuel Type and Color. Describe how you would convert these to binary variables. Confirm this using __pandas__ methods to transform categorical data into dummies. What would you do with the variable _Model_?

__Answer:__
The categorical fuel_type variable has three categories: petrol, diesel and CNG. To convert these variables into dummy variables, we use only two variables (here we use Petrol and Diesel, but a different pair can also be chosen). The binary variable Petrol gets the value 1 if Fuel Type=Petrol and otherwise it gets the value 0. The binary variable Diesel gets the value 1 if Fuel Type=Diesel and otherwise it gets the value 0. CNG is the "reference category." If Fuel type is CNG, both binary variables take the value 0.

Similarly, the variable _Color_ is converted to dummy variables. It resulted in 10 dummy variables with "Color_Beige" as the reference category.

Making dummies of all the categories in _Model_ would make for a lot of predictor variables.  Some preliminary exploration should be done first, to see if there are a small number of models that account for most cases.  The analysis could be confined initially to those to see how important Model is as a predictor.

In [7]:
# load data
toyota_df = pd.read_csv(DATA / 'ToyotaCorolla.csv')
# data dimension
print("\n(#Rows, #Columns):", toyota_df.shape)
print("\n")
# review first few records
print("First few records:\n")
print(toyota_df.head())
# create dummy variables for categorical variables, ignore the variable Model
toyota_df = pd.get_dummies(toyota_df.iloc[:,3:39], prefix_sep='_', drop_first=True)
# print column/variable names
print("\nVariables:", toyota_df.columns)
# review first few records in dummy variables
print("\nFirst few records in dummy variables:")
print(toyota_df.loc[:, 'Fuel_Type_Diesel':'Color_Yellow'].head(5))


(#Rows, #Columns): (1436, 39)


First few records:

   Id                                          Model  Price  Age_08_04  \
0   1  TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13500         23   
1   2  TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13750         23   
2   3  TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13950         24   
3   4  TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  14950         26   
4   5    TOYOTA Corolla 2.0 D4D HATCHB SOL 2/3-Doors  13750         30   

   Mfg_Month  Mfg_Year     KM Fuel_Type  HP  Met_Color  ... Powered_Windows  \
0         10      2002  46986    Diesel  90          1  ...               1   
1         10      2002  72937    Diesel  90          1  ...               0   
2          9      2002  41711    Diesel  90          1  ...               0   
3          7      2002  48000    Diesel  90          0  ...               0   
4          3      2002  38500    Diesel  90          0  ...               1   

   Power_Steering  Radio  M

__2.11.b.__ Prepare the dataset (as factored into dummies) for data mining techniques of supervised learning by creating partitions in Python. Select all the variables and use default values for the random seed and partitioning percentages for training (50%),
validation (30%), and test (20%) sets. Describe the roles that these partitions will play in modeling.

__Answer__


In [8]:
# partition the data into training(50%), validation (30%) and test(20%) sets
# random_state is set to a defined value to get the same partitions when re-running the code
# training (50%), validation (30%), and test (20%)
from sklearn.model_selection import train_test_split

trainData, temp = train_test_split(toyota_df, test_size=0.5, random_state=1)
validData, testData = train_test_split(temp, test_size=0.4, random_state=1)
print('Training : ', trainData.shape)
print('Validation : ', validData.shape)
print('Test : ', testData.shape)

Training :  (718, 45)
Validation :  (430, 45)
Test :  (288, 45)


__Training dataset__

The training dataset is used to train or build models. For example, in a linear regression, the training dataset is used to fit the linear regression model, i.e. to compute the regression coefficients. This is usually the largest partition.

__Validation dataset__

Once a model is built on training data, we assess the accuracy of the model on unseen data. For this, the model should be used on a dataset that was not used in the training process. In the validation data we know the actual value of the response variable, and can therefore examine the difference between the actual value and the predicted value to determine the error in prediction. Based on this performance, sometimes the validation dataset is used to tweak the model, or to choose between multiple fitted models.

__Test dataset__

The validation dataset is often used to select a model with minimum error. Testing that model on completely unseen data gives a realistic estimate of the performance of the model. When a model is finally chosen, its accuracy with the validation dataset is still an optimistic estimate of how it would perform with unseen data. This is because (1) the final model has come out as the winner among the competing models based on the fact that its accuracy with the validation dataset is highest, and/or (2) the validation set was used to help build one or more models. Thus, you need to set aside yet another portion of data, which is used neither in training nor in validation, which is called the test dataset. The accuracy of the model on the test data gives a realistic estimate of the performance of the model on completely unseen data.

__*It has been pointed out that there is a value in the cc variable - 16,000 - that is probably a data input error. The solutions have been prepared without correcting this error, but a solution that includes correcting this error to 1600 would also be fine. (The data could also be used as a small illustration or exercise of data prep and cleaning.)__